More FCI fixes

bendudson opened the pull request:
Working through normalisation and operators

- With FCI grids J has units of volume, so should be divided by rho_s0^3 to normalise.
- In `vorticity` and `relax_potential`, curvature vector X component needs to be normalised differently in FCI coordinates.
- Add `exb_advection` option to disable ExB advection terms
- Add `bracket_factor` to ExB poisson bracket terms

bendudson committed:
FCI metric: Fix normalisation of J

With FCI grids J has units of volume, so should be divided
by rho_s0^3 to normalise.

bendudson committed:
Div_par_mod, Div_par_fvv: Modify FCI implementations

When using FCI can't average metrics across poloidal planes.

bendudson committed:
ion_viscosity: Use Div_par_K_Grad_par_mod

Same operator as parallel thermal diffusion in evolve_pressure

bendudson committed:
sheath_boundary_parallel: Use B rather than J

J can't have yup/down parallel slices, but B can.

bendudson committed:
Div_par_fvv: Add penalty term to FCI version

Include a term proportional to local maximum wave speed, that penalises
jumps across cells. For now using the lowest order, most diffusive,
form of this term.

bendudson committed:
Calculate Bxy yup/down by interpolation

bendudson committed:
div_ops: New Div_par_mod and Div_par_fvv for FCI

Include dissipation terms using amax local wave speed.

bendudson committed:
vorticity: Change curvature normalisation for FCI

When using an FCI mesh the curvature vector components
have different dimensions to the usual Clebsch coordinates.

bendudson committed:
div_ops: Replace Div_par_K_Grad_par_mod for FCI

Make method similar to that used in field-aligned coordinates.

bendudson committed:
sheath_boundary_parallel: Revert changes to using B

Return to using g_22 and J, if metric parallel slices are
calculated by Zoidberg.

bendudson committed:
evolve_momentum: NV slices handled automatically

No need to set yup/down slices manually, since the arithmetic
operators handle this.

bendudson committed:
ExB advection: Fixes for FCI meshes

- Add an input option `exb_advection` to disable ExB terms.
  Default is true (ExB advection included) for backward compatibility.

- Calculate `bracket_factor` and apply to Poisson bracket terms.
  This factor should probably be calculated in Coordinates.

bendudson committed:
relax_potential: Normalise curvature for FCI

Curvature vector needs to be normalised differently in FCI
coordinates because x is not a poloidal flux.

bendudson committed:
relax_potential: Suppress log output, set phi boundaries

- Use get<>() to get Tnorm, so it isn't logged
- Set phi parallel boundary

bendudson committed:
evolve_momentum: Include parallel gradients of phi

If exb_advection = false, was removing all use of phi including
parallel electric field. Now only disables ExB terms.

bendudson committed:
sheath_boundary_parallel: Add always_zero_current

Option to set potential based on zero current through the
boundary, even if phi is set by e.g. vorticity.

bendudson committed:
BOUT++: Update external dependency

bendudson committed:
Update BOUT++ version

Fixed typo

bendudson committed:
relax_potential: Add viscosity, phi_dissipation, bracket_factor

- viscosity is a kinematic viscosity, diffusing vorticity

- phi_dissipation is a numerical suppression of potential variation
  along magnetic fields, using fastest wave speed.

- bracket_factor is needed on ExB terms in non-Clebsch coordinate systems.

bendudson committed:
diamagnetic_drift: Fix curvature normalisation for FCI

- If curvature isn't read, throw an exception rather than silently
  continuing with no curvature
- If using FCI, curvature normalisation needs to be different,
  because x is dimensionless (rather than a flux).

bendudson committed:
Update BOUT++

bendudson committed:
evolve_pressure: Set kappa boundaries for FCI

bshanahan commented:
There seems to be a change with X-splitting somewhere. Turbulence simulations are stable-ish, but I see the processor boundaries in vorticity:

![image](https://github.com/user-attachments/assets/c6c16656-5b8d-4bff-9f09-c1a735cec5c8)

bshanahan committed:
Communicate density in fixed_density if using FCI

bshanahan committed:
Merge branch 'fci-auto-with-debug' into fci-more-fixes

totork commented:
On the topic of normalisation:
Do we also have to normalize coords->G1 and coords->G3, they are used for the laplace inversion when all_terms=true @bendudson  @dschwoerer

dschwoerer commented:
No, because we recalculate them [here](https://github.com/dschwoerer/hermes-3/pull/5/files#diff-2194b1467977c9d2bc25c3908b67a4c3fdfb245fbf5bdf71c090bd01a99c030eL297)

totork commented:
A point that could possibly cause stability issues is the use of `dy.ydown()[iym] `and `dy.yup()[iyp]` in combination with an issue in zoidberg:

https://github.com/boutproject/zoidberg/issues/39

Maybe it is worth checking if this is also coming up in your grids. If it does, that could cause weird numerical effects. It should e.g. appear in the heat conduction operator 

`gradient = 2. * (fin[i] - f_down[iym]) / (coord->dy[i] + coord->dy.ydown()[iym]);`

with commit 37aff96c491e8426a9cb0edfc47de3eeef367d49

@dschwoerer @bshanahan Maybe it is worth checking

dschwoerer commented:
That sounds to me like a bug in zoidberg. dy should be positive. Maybe 
add a check in bout to ensure the grid isnt buggy. I will try to have a 
look next week. @totork Could you open a zoidberg bug?

On Wed, May 28 2025 at 07:35:52 -07:00:00, totork 
***@***.***> wrote:
> *totork*
>  left a comment
> (dschwoerer/hermes-3#5) 
> <https://github.com/dschwoerer/hermes-3/pull/5#issuecomment-2916587335>
> A point that could possibly cause stability issues is the use of 
> dy.ydown(iym)and dy.yup(iyp) in combination with an issue in zoidberg:
> 
> boutproject/zoidberg#39 
> <https://github.com/boutproject/zoidberg/issues/39>
> 
> Maybe it is worth checking if this is also coming up in your grids. 
> If it does, that could cause weird numerical effects. It should e.g. 
> appear in the heat conduction operator
> 
> g_22 = 0.5 * (coord->g_22[i] + coord->g_22.ydown()[iym]);
> 
> with commit 37aff96 
> <https://github.com/dschwoerer/hermes-3/commit/37aff96c491e8426a9cb0edfc47de3eeef367d49>
> 
> @dschwoerer <https://github.com/dschwoerer> @bshanahan 
> <https://github.com/bshanahan> Maybe it is worth checking
> 
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/dschwoerer/hermes-3/pull/5#issuecomment-2916587335>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABLAMHS4VYNALYUKLWIWVG33AXCURAVCNFSM6AAAAABZBWDGFKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMJWGU4DOMZTGU>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

totork commented:
One more question on the relax_potential:

We are using 

`ddt(phi1) =  lambda_1 * (Div_a_Grad_perp(average_atomic_mass / Bsq, phi) - Vort);`

This does not use the dagp operator for fci? 
@dschwoerer

bendudson committed:
isothermal: Communicate T if FCI

Need to fill yup/down fields

bendudson committed:
isothermal: Fix typos, make T a Field3D

For FCI simulations T needs yup/ydown fields.
When a BoutReal is converted to a Field3D the yup/down fields
are not set.

bendudson committed:
FCI: Reverting changes to Div_par_mod

Seem to cause problems with inner boundary: Observe
temperature falling to low values from the inner (X) boundary.

totork committed:
Fix ion viscosity boundary

totork committed:
Added fci version of diamagnetic drift

totork committed:
Added bracket for of diamagnetic drift in vorticity

totork committed:
Changed default behavior of vorticity to use bracket in FCI

totork committed:
Added fci blob2d to examples

totork committed:
Added example plots

totork committed:
Added readme to fci blob example

totork committed:
Fix BC for kappa in evolve pressure

totork committed:
Added unit tests to diamagnetic drift fci

totork committed:
Added another unit test for diamagnetic_fci

totork committed:
More unit tests for diamagnetic_fci

totork committed:
Changed electron viscosity boundary conditions

totork committed:
Changed BC for ion viscosity

dschwoerer committed:
Update BOUT++

totork committed:
Merge branch 'fci-more-fixes' into fci-dia

totork committed:
Added adhoc potential

totork committed:
First start of fluid mms test

totork committed:
Added check to not doublecount source

totork committed:
Use more accurate div_par operator with less dissipation

totork committed:
Add flags for mms tests

totork committed:
Add flags for MMS test in pressure

totork committed:
Added flag for mms tests in momentum

totork committed:
Added MMS test for density in fci

totork committed:
Added MMS test for pressure in fci, but without conduction

totork committed:
Change CMake file for the tests

totork committed:
Added normal div_par in momentum for second order MMS convergence test

totork committed:
Added integrated tests to cmake

totork committed:
updated MMS test for 1d-fluid-fci

totork committed:
1D conduction mms test for fci

totork committed:
Fixed electron viscosity and added MMS test for this

totork committed:
Added ion viscosity mms test

totork committed:
Added momentum mms test

totork committed:
First commit to fix relax potential

totork committed:
Change operator of div_par

totork committed:
Added dissipative option for parallel divergence

totork committed:
Added separate collision times for self collisions, added dissipative option for parallel divergence

bendudson committed:
isothermal: Fix missing mesh

bendudson committed:
test_diamagnetic_drift_fci: Add 3D metric guard

Test works for 3D metrics

totork committed:
Added label for unit tests

totork committed:
Added way of not extrapolating into sheath but using neumann

totork committed:
Added grids for mms tests as tar file

totork committed:
Adjusted tests for the grids

totork committed:
Added new perpendicular diffusion operator

totork committed:
Added new diffusion operator to anomalous_diffusion_3d

totork committed:
Added 2D diffusion test

totork committed:
Change repo of grids for tests

totork committed:
Decrease number of grids for 2D anomalous diffusion test

bendudson committed:
Merge pull request #418 from totork/fci-dia

Diamagnetic drift for FCI + bugfixes

totork committed:
Added tar file to 2D mms test

bendudson committed:
Merge pull request #419 from boutproject/fci-more-fixes-update

Update BOUT++ in FCI branch

totork committed:
Changed to spatially dependent diffusion coefficients in 2D diffusion test

totork committed:
Removed setting neumann on T, otherwise incostent when setting dirichlet at mms tests

totork committed:
Added perp diffusion test for density and pressure

totork committed:
Fix typo in ddt(phi1)

totork committed:
Reverted back to use the complete collision frequency for conduction/viscosity

totork committed:
Added 2D tests to hermes test suite

totork committed:
Bugfix for fastest wave factor

bendudson committed:
Div_par_fvv: FCI change numerical dissipation

Damp gradients of velocity rather than gradients of momentum.

bendudson committed:
Div_par_fvv: FCI change numerical dissipation

Damp gradients of velocity rather than gradients of momentum.

totork committed:
Changed dissipation in operator of parallel momentum advection

totork committed:
Allow for different boundary conditions for phi in relaxation

bendudson committed:
evolve_pressure: Communicate T, apply boundaries

With FCI, need to communicate and interpolate T in addition to
interpolating P and N.

totork committed:
Output velocities

totork committed:
Added scale for ExB advection to make start of simulation easier

dschwoerer committed:
Ensure we do not compute parallel slices

dschwoerer committed:
ensure we apply parallel BCs after communication

dschwoerer committed:
re-enable copying examples and tests

dschwoerer committed:
Add fci-example-1 test

dschwoerer committed:
Fix pip install script

dschwoerer committed:
CI: Speedup build

dschwoerer committed:
Add extra target for setup-test

dschwoerer committed:
Fix download test data

dschwoerer committed:
update BOUT++

dschwoerer committed:
Switch to launch_safe

dschwoerer committed:
New fvv is less smoothing but more stable

dschwoerer committed:
Fixup bad cmake merge

dschwoerer committed:
Cleanup runtest

dschwoerer committed:
update BOUT++

dschwoerer committed:
fix CONFLICT declaration

dschwoerer committed:
prefer launch_safe

dschwoerer committed:
Remove debugging print statements

dschwoerer committed:
Add conflicts to tests

Damp gradients of velocity rather than gradients of momentum.

dschwoerer committed:
update BOUT++

dschwoerer committed:
Some tests do not work yet with 3D metrics

dschwoerer committed:
Fix hermes-3 binary path

Also, there is no need to add a symlink

dschwoerer committed:
Fix unit test for 2D metrics

dschwoerer committed:
Switch to default solver

dschwoerer committed:
Skip SNB unit test for 3D metrics

dschwoerer committed:
Run FCI only with 3D metrics

dschwoerer committed:
Fix unit test

totork committed:
Merge pull request #2 from boutproject/toto-relax-test

Working on tests

totork committed:
Added manual fix for grids in fci

totork committed:
Explicitly set BCs

totork committed:
BCs in neutral model

totork committed:
Fixes for FCI in the AFN

totork committed:
Added Bufferand kmodel

totork committed:
Added kmodel to get diffusion coefficients

totork committed:
Output source terms once for postprocessing

totork committed:
More FCI specific fixes for the neutral mixed model

totork committed:
More diagnostic output

totork committed:
Added hyperdiffusion to vorticity

totork committed:
Add communication to vorticity

totork committed:
Added optional heating

totork committed:
Added option to artifically increase momentum and energy exchange

totork committed:
Add option to disable time evolution of vort and phi

totork committed:
Merge branch 'fci-kmodel' into fci-adhoc

totork committed:
Merge pull request #5 from totork/fci-adhoc

Added adhoc potential

totork committed:
Fixed adhoc potential

totork committed:
Merge pull request #6 from totork/fci-neutrals

Merge FCI neutral changes

totork committed:
not multiplying unused fields

totork committed:
Add hyperdiffusion to density, momentum and pressure

totork committed:
Separate output of ddt

totork committed:
Add output for ddt that is separate

totork committed:
Add adaptive source terms

totork committed:
Fix error in TE of neutral afn model

totork committed:
Change agressiveness of adaptive source in density

totork committed:
Fixed electromagnetic BC application

totork committed:
Add parallel diffusion for vorticity

totork committed:
More changes for trace denssity

totork committed:
Add adaptive source to keep density above certain value without enforcing flooring directly

totork committed:
Fix bug that would cause a not initialised bug

totork committed:
Merge pull request #7 from totork/fci-trace

Add trace density for fci and add local sources

totork committed:
Add finite difference for neutral mixed model

totork committed:
Use more dissipative parallel advection

totork committed:
First commit of recycling fci component

totork committed:
Fix random setting of vnnlim in neutral model

totork committed:
Adapted ionisation

totork committed:
Added tracedensty evolution

totork committed:
Fix bug operation on empty data

totork committed:
Make neutral_lmax not hardcoded but as input option

totork committed:
Fix typo in recycling

totork committed:
Changed neutral model preconditioner

totork committed:
Changes to fit neutral_mixed for mms tests

totork committed:
2D mms test of perpendicular density and pressure diffusion for fci's AFN

totork committed:
Fix prefactor in AFN conduction

totork committed:
Add 2D MMS test for AFN including conduction but still without momoentum

totork committed:
Revert change, otherwise parallel slices are not set

totork committed:
Disable sound speed dissipation when in MMS

totork committed:
Fix FCI version of the AFN

totork committed:
Revert "Fix FCI version of the AFN"

This reverts commit afff9e7ca715c1fc16d9934169d8a25413fc5a14.

totork committed:
Fix parallel pressure advection in AFN

totork committed:
Add option for mms tests to use exact operators without numerical diffusion

totork committed:
Add 1D MMS test for AFN

totork committed:
Added slab grids with BCs in parallel direction

totork committed:
Get rid of mass dependence of neutral viscosity in AFN

totork committed:
Change 1D neutral mms test

totork committed:
Add second row of cells that see the boudary in parallel direction

totork committed:
Only apply BCs in AFN when not in MMS test case

totork committed:
Only apply BCs when not an MMS test

totork committed:
Fix neutral 1d test of AFN

totork committed:
Fix CMakeLists to include neutral mms tests

totork committed:
Added test for preconditioner in neutral model

totork committed:
Integrated test for neutral preconditioner

totork committed:
Added vorticity loss term

totork committed:
Add grid-dependend dissipation of vorticity to localize dissipation in the core

dschwoerer committed:
CI: Build with PETSc

totork committed:
Disable 2D neutral preconditioner test

dschwoerer committed:
Fix recycling - sqrt got lost in rewrite

totork committed:
Merge pull request #9 from boutproject/toto-fci-kmodel

Fix recycling

totork committed:
Fix sign mistake in the neutral model

totork committed:
Added 2D MMS test for diffusion: dagp operator not converging but FD does

totork committed:
Change order of MMS test

totork committed:
Output successfull extraction of grid

totork committed:
Change from MagneticField class to Slab class in circular grid

totork committed:
Fix linking of test

totork committed:
Stop linking completely in MMS test

totork committed:
Launch_safe in MMS test

totork committed:
Change to pvode in MMS

totork committed:
CI: add sundials in 3D metrics

totork committed:
Fix actual calculation of recycling flux to align it to the operator

totork committed:
Added option neumann BCs for neutral model, consistent with how GRILLIX does their neutral model in FCI

totork committed:
Add script to generate straight circular grid

totork committed:
Clean up code

bendudson committed:
Merge remote-tracking branch 'totork/fci-kmodel' into fci-more-fixes
